src/guide/typescript-support.md

```diff
@@ -133,6 +133,45 @@ const Component = defineComponent({
 })
 ```
 
+### Augmenting Types for Use with Plugins or Declarations by User
+
+Vue 3 provides `globalProperties` property for [plugins](./plugins.html#writing-a-plugin) to add to Vue’s global properties in each component instance. In some cases, we also need to [do it by ourselves](../api/application-config.html#globalproperties):
+
+```typescript
+// User Definition
+import axios from 'axios'
+
+const app = Vue.createApp({})
+app.config.globalProperties.$http = axios
+
+// Plugin for validating some data
+export default {
+  install(app, options) {
+    app.config.globalProperties.$validate = (data: object, rule: object) => {
+      // check whether the object meets certain rules
+    }
+  }
+}
+```
+
+In order to let TypeScript know the properties we (or plugins) added and their types, the [Module Augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation) feature can help us.
+
+For the above example, we (or plugins) can add the following code of type definition:
+
+```typescript
+import axios from 'axios'
+
+declare module '@vue/runtime-core' {
+  export interface ComponentCustomProperties {
+    axios: typeof axios
+  }
+}
+```
+
+We should put the above code wherever it can be loaded, such as `main.ts` where you define these properties, or a `*.d.ts` file in `src/typings` folder automatically loaded by TypeScript. For plugins, it should be placed in the location specified by the `types` property in `package.json`. 
```

Comment:
```suggestion
We can put this type declaration in the same file, or in a project-wide `*.d.ts` file (for example, in the `src/typings` folder that is automatically loaded by TypeScript). For library/plugin authors, this file should be specified in the `types` property in `package.json`.
```

Reply:
Adding this to a .d.ts file is not a good idea in my experience, since that will effectively re-declare the runtime-core modules which means all the types that this module exports will be no longer accessible.

Reply:
That is only the case if nothing is imported into the module declaration file, or if the user attempts to extend a type rather than an interface.

Reply:
According to TypeScript docs, the declarations in an augmentation will be **merged** rather than **replaced**, whether it is declared in `.ts` or `.d.ts` file.

From another perspective, the developer of the library/plugin can only generate the type declaration to a separate `.d.ts` file (because the final release is the JS file packaged by the rollup), such as [the Vue-Router Type declaration file (Please see the end of the file)](https://www.unpkg.com/vue-router@4.0.5/dist/vue-router.d.ts).

Reply:
@dev-itsheng If there is no other import in the file then it is not treated as a module and the behaviour that @LinusBorg describes occurs.

Reply:
Yes, I just tried some cases. Indeed, if only use `declare module` in `d.ts` (and `.ts`) files, the declaration will be overwritten instead of merged. When I add an `import` (no matter what I import), it will return to normal.

Then I checked the TypeScript documentation. It did not describe my scenario, but mentioned in [another place](https://www.typescriptlang.org/docs/handbook/modules.html) that files with top-level import/export would be regarded as modules, otherwise they would be scripts. I guess TypeScript might be able to deal with these two Different file types are handled differently.

So, I used `export` instead of `import` (such as `export const a: number` or `export const a = 1`) and found that it still works.

In this case, I think it is necessary to explain it in the document, too.